src/shared/image-policy.c

```diff
@@ -0,0 +1,653 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "extract-word.h"
+#include "image-policy.h"
+#include "sort-util.h"
+#include "string-util.h"
+#include "strv.h"
+
+/* Rationale for the chosen syntax:
+ *
+ * → one line, so that it can be reasonably added to a shell command line, for example via `systemd-dissect
+ *   --image-policy=…` or to the kernel command line via `systemd.image_policy=`.
+ *
+ * → no use of "," or ";" as separators, so that it can be included in mount/fstab-style option strings and
+ *   doesn't require escaping. Instead, separators are ":", "=", "+" which should be fine both in shell
+ *   command lines and in mount/fstab style option strings.
+ */
+
+static int partition_policy_compare(const PartitionPolicy *a, const PartitionPolicy *b) {
+        return CMP(ASSERT_PTR(a)->designator, ASSERT_PTR(b)->designator);
+}
+
+static PartitionPolicy* image_policy_bsearch(const ImagePolicy *policy, PartitionDesignator designator) {
+        if (!policy)
+                return NULL;
+
+        return typesafe_bsearch(
+                        &(PartitionPolicy) { .designator = designator },
+                        ASSERT_PTR(policy)->policies,
+                        ASSERT_PTR(policy)->n_policies,
+                        partition_policy_compare);
+}
+
+static PartitionPolicyFlags partition_policy_normalized_flags(const PartitionPolicy *policy) {
+        PartitionPolicyFlags flags = ASSERT_PTR(policy)->flags;
+
+        /* If no protection flag is set, then this means all are set */
+        if ((flags & _PARTITION_POLICY_USE_MASK) == 0)
+                flags |= PARTITION_POLICY_OPEN;
```

Comment:
This is actually what happens. But on another level.

So if you define a policy and leave out a partition so that its policy is unspecified we'll actually read that restrictively: the implied policy for that partition is then "ignore". so this is what you are asking for!

But if you *do* specify policy for a partition, but don't specify any of the "use" flags for it (i.e. the flags that say "ignore", "use without protection", "require encryption", "require verity"), then we'll imply "use in any mode". why that? my thinking there is: you apparently care about the partition type otherwise you would have omitted it. But you don't care how precisely it is used (otherwise you'd have specified that), hence "open" is a good idea, i.e. without any restrictions. This actually helps keeping policy expressions short, too.

or in other words: if the default for undeclared partition types is to ignore, and you do declare the partition anyway, than you probably have something else in mind than "ignore", hence just make use of it.

Moreover this behaviour matches the behaviour for the read-only-flag-on and read-only-flag-off bits: if neither of these flags are set we also read it as "can be anything", i.e. as if both bits are on.